oak_channel/SPEC.md

```diff
@@ -138,149 +93,80 @@ A byte encoded frame MUST consist of the following fields:
-"padding:32,length:16,flags:16,body bytes...:64" -->
+<!-- Diagram generated with https://www.luismg.com/protocol/, using the spec
+"invocation_id:32,message_length:32,frame_number:32,body_length:32,body bytes...:64" -->
 
-- `padding`, 4 bytes
+- `invocation_id`, u32, little endian
 
-  For future use.
+  The unique ID of the invocation that the frame's forms a part of. All the
+  frames that make up a message MUST have the same invocation ID. The invocation
+  ID of a response message MUST match the value of its related request message.
+  The invocation ID MUST be incremented for each method invocation, naturally
+  wrapping when the maximum value for the data type is reached.
 
-- `length`, u16, little endian
+- `message_length`, unsigned 32-bit little-endian integer
 
-  Length of the entire frame, including this field. The length of a frame MUST
-  NOT exceed 4,000 bytes.
+  The length of the overall message. This value MUST be larger or equal to
+  `body_length`. The message length for all frames that make up the message MUST
+  be the same.
 
-- `flags`, 2 bytes
+  The size of a single message MUST NOT exceed 2GiB.
 
-  - `flags[0]`
+- `frame_number`, unsigned 32-bit little-endian integer
 
-    - 7 (MSB): unused
-    - 6: unused
-    - 5: unused
-    - 4: unused
-    - 3: unused
-    - 2: unused
-    - 1: end flag. MUST be set in the last frame of a message
-    - 0 (LSB): start flag. MUST be set in the first frame of a message
+  The position of this frame within the overall message. The first frame MUST
+  start at number 0. Each frame's number MUST be 1 higher than the previous
+  frame. Frames for a single message MUST be sent in ascending order of the
+  frame number.
 
-  - `flags[1]`
+- `body_length`, unsigned 32-bit little-endian integer
 
-    - 7 (MSB): unused
-    - 6: unused
-    - 5: unused
-    - 4: unused
-    - 3: unused
-    - 2: unused
-    - 1: unused
-    - 0 (LSB): unused
+  The length of the body byte array of the rfram. This value must not exceed
+  4,080. The sum of the body lengths of all the frames in a message must equal
+  the message length.
 
 - `body`, variable length byte array
 
-  Byte fragment of an encoded message. In order to not exceed the maximum frame
-  length, this byte array MUST NOT exceed 3,936 bytes.
+  Byte fragment of an encoded message. This size of the byte array MUST NOT
+  exceed 4,080 bytes.
+
+The maximum total length of a single frame is 4,096 bytes.
 
 ## Interaction between the Layers
 
 ### Sending Messages
 
-To send a message, it MUST first be encoded into frames. The resulting frame set
-MUST be sent as an uninterrupted sequence.
-
-### Receiving Messages
-
-To receive a message the recipent incrementally reads frames from the
-communication channel. The recipent MAY start parsing a message before all of
-its frames have been received.
-
-### Encoding Messages into Frames
+To send a message, it MUST first be encoded into frames.
```

Comment:
nit: Can we formulate to active? Is it the "Sender" who has to do this?

To send a message, the sender MUST encode a message into frames. The sender MUST chunk the byte representation of a message into a set of frame bodies. [..]

The sender must form a frame from each frame body and MUST send the resulting frame set in sequenctial order of the frame number over the channel.

Reply:
+1

Reply:
Done.